working/0546-patterns/patterns.md

```diff
@@ -0,0 +1,924 @@
+
+Determines if the value is equivalent to the value of the given constant
+expression. The expression is syntactically restricted to be either:
+
+*   **A bare identifier.** In this case, the identifier must resolve to a
+    constant declaration in scope.
+
+*   **A prefixed or qualified identifier.** In other words, `a.b`. It must
+    resolve to either a top level constant imported from a library with a
+    prefix, a static constant, or an enum value.
+
+*   **A prefixed qualified identifier.** Like `a.B.c`. It must resolve to a
+    value on an emum type that was imported with a prefix, or a static
+    declaration in a class, mixin, or extension imported with a prefix.
+
+The const expression is evaluated and the result passed to the matched value's
+`==` operator to determine if it matches.
+
+**TODO: Do we want to allow other kinds of constant expressions like `1 + 2`?
+Switch statements allow arbitrary const expressions. We probably don't want to
+allow const collection literals since those clash with the corresponding
+collection patterns.**
+
+### List matcher
+
+```
+listMatcher ::= "[" matchers "]"
+```
+
+Matches and destructures lists. If the value's type does not implement `List<T>`
+for some T, then the match fails. If the length of the list determined by
+calling `length` is not equal to the number of subpatterns, the match fails.
+
+Otherwise, extracts a series of elements from the list using the index operator
+and tests them against the corresponding subpatterns. The list matcher succeeds
+if all of the subpatterns match.
+
+**TODO: Allow a `...` element to skip over elements and match suffixes. Can you
+capture the rest in a variable?**
+
+**TODO: Should we loosen this to match against any Iterables by using
+`elementAt()`?**
+
+**TODO: Can the pattern have a type argument? If so, can that have a type
+pattern?**
+
+### Map matcher
+
+```
+mapMatcher ::= "{" mapMatcherEntries "}"
+
+mapMatcherEntries ::= mapMatcherEntry ( "," mapMatcherEntry )* ","?
+mapMatcherEntry ::= expression ":" matcher
+```
+
+Matches and destructures maps. If the value's type does not implement `Map<K,V>`
+for some K and V, then the match fails. Otherwise, tests the entry patterns:
+
+1.  For each `mapMatcherEntry`:
+
+    1.  Evaluate `expression` and pass to `containsKey()` on the value.
+
+    1.  If this returns `false`, the map does not match.
+
+    1.  Otherwise, pass the expression's value to the value's index operator.
+        Match the entry's matcher against the resulting value. If it does not
+        match, the map does not match.
+
+1.  If all entries match, the map matches.
+
+Note that, unlike with lists, a matched map may have additional entries that
+are not checked by the matcher.
+
+If it is a compile-time error if any of the entry key expressions are not
+constant expressions.
+
+**TODO: Can the pattern have type arguments? If so, can they have type
+patterns?**
+
+### Typed matcher
+
+```
+typedMatcher ::= type identifier
```

Comment:
This is going to be confusing, at least at first.  I think we'll have users who write `int x` instead of `var x`, and are surprised when the match fails.